app/models/user.rb

```diff
@@ -136,8 +141,20 @@ def recent_active_offer_id
     Version.for_offers.by_user(id).last.try(:related_id_or_item_id)
   end
 
+  def create_or_remove_user_roles(role_ids)
+    remove_user_roles(role_ids)
+    role_ids.each do |role_id|
+      user_roles.where(role_id: role_id).first_or_create
+    end
+  end
+
   private
 
+  def remove_user_roles(role_ids)
```

Comment:
We don't appear to be using this method anywhere else apart from in `create_or_remove_user_roles` so I would be tempted to merge this into `create_or_remove_user_roles` and change the method name to `set_user_roles`. This is because when I read `remove_user_roles(role_ids)` I think that `role_ids` is the list of roles to remove from a user but actually that is not the case here, its the list of roles to keep! E.g. if `role_ids` is `nil` then it will remove *all* roles rather than none.

Also, shouldn't `roles.pluck(:id) - role_ids` be the other way around? To get the list of roles to remove.

```ruby
def set_user_roles(role_ids)
  role_ids_to_remove =  role_ids - roles.pluck(:id)
  user_roles.where(role_id: role_ids_to_remove).destroy_all
  role_ids.each do |role_id|
    user_roles.where(role_id: role_id).first_or_create
  end
end
```

Reply:
good catch for `not in []` ! as long as it is well tested, this will work